entropy/coin_flip/app/src/flip_coin.ts

```diff
@@ -29,75 +45,80 @@ const parser = yargs(hideBin(process.argv))
 
 async function main() {
   const argv = await parser.argv;
+  const chainName = argv.chainName;
+  if (chainName !== "optimism-sepolia" && chainName !== "arbitrum-sepolia") {
+    throw new Error("Invalid chain name");
+  }
+
+  let chain: Chain;
+  if (chainName === "optimism-sepolia") {
+    chain = optimismSepolia;
+  } else {
+    chain = arbitrumSepolia;
+  }
+
+  const client = createWalletClient({
+    chain: chain,
+    account: privateKeyToAccount(argv.privateKey as Hex),
+    transport: http(argv.rpcUrl),
+  }).extend(publicActions);
+
+  const coinFlipContract = getContract({
+    address: argv.address as Hex,
+    abi: ICoinFlipAbi,
+    client: client,
+  });
+
+  console.log("1. Generating user's random number...");
 
-  const coinFlipContractAddress = argv.address;
-  const rpc = argv.rpcUrl;
-  const privateKey = argv.privateKey;
+  const randomNumber = `0x${crypto.randomBytes(32).toString("hex")}`;
+  console.log(`User Generated Random number: ${randomNumber}`);
 
-  const provider = new HDWalletProvider({
-    privateKeys: [privateKey],
-    providerOrUrl: rpc,
+  console.log("\n2. Requesting coin flip...");
+
+  const flipFee = await coinFlipContract.read.getFlipFee();
+  console.log(`Flip Fee: ${flipFee} wei`);
+
+  console.log("\n3. Sending request to flip coin...");
+
+  let sequenceNumber: bigint;
+  coinFlipContract.watchEvent.FlipRequest({
+    onLogs: (logs) =>
+      logs.forEach((log) => {
+        console.log(`Flip Request Number/ Sequence Number: ${log.args.sequenceNumber}`);
+        sequenceNumber = log.args.sequenceNumber as bigint;
+      }),
   });
 
-  const web3 = new Web3(provider as any);
+ await client.watchBlocks({
+    onBlock: (block) => {
+      console.log(`\nPolling Block: ${block.number}`);
+      coinFlipContract.watchEvent.FlipResult({
+        onLogs: (logs) =>
+          logs.forEach((log) => {
+            if (log.args.sequenceNumber === sequenceNumber) {
+              console.log(
+                `\nFlip Result: ${log.args.isHeads ? "Heads" : "Tails"}`
+              );
+              process.exit(0);
+            }
+          }),
+      });
+    },
+  });
```

Comment:
I'm no viem expert but this seems like a really inefficient way to subscribe to the event, if I'm understanding correctly this will create a new subscription to the `FlipResult` for each block?  Is there any actual reason to be doing all this inside a `watchBlocks` callback?

Again I don't know viem well but naiively I'd expect the flow to be something like this:

1. Generate random number
2. Get flip fee
3. Request a flip
4. Get the transaction hash from the flip request (viem returns this)
5. Get the sequence number from the transaction hash
6. Watch for `FlipResult` matching the sequence number

I'm not sure why you'd need to watch for events to get the sequence number, or why you'd have to watch blocks at all.

Reply:
hmm there's no way to get the event from the hash?  That's a shame, this feels messier than I would expect....